pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,648 @@
+		for _, removeBk := range backends {
+			if removeBk.StringWithProtocol() != currBk.StringWithProtocol() {
+				newBackends = append(newBackends, currBk)
+			}
+		}
+	}
+	return newBackends
+}
+
+func (rpm *Manager) deletePolicyBackends(config *LRPConfig, backends ...backend) {
+	// Currently, we expect number of LRP backends to be a single digit number.
+	// If this scales up, we might need to optimize this using sets.
+	if config.lrpType == lrpConfigTypeAddr {
+		config.frontendMappingForAddr.backends = rpm.filterBackends(config.frontendMappingForAddr, backends...)
+		rpm.notifyPolicyBackendDelete(config, config.frontendMappingForAddr)
+	} else {
+		for _, fe := range config.frontendMappingForSvc {
+			fe.backends = rpm.filterBackends(fe, backends...)
+			rpm.notifyPolicyBackendDelete(config, fe)
+		}
+	}
+}
+
+// Deletes service entry for the specified frontend.
+func (rpm *Manager) deletePolicyFrontend(config *LRPConfig, frontend frontend) {
+	found, err := rpm.svcManager.DeleteService(frontend)
+	delete(rpm.frontendsByHash, frontend.Hash())
+	if !found || err != nil {
+		log.WithError(err).Debugf("Local redirect service for policy %s not deleted",
+			config.id)
+	}
+}
+
+// Updates service manager with the deleted backend(s).
+func (rpm *Manager) notifyPolicyBackendDelete(config *LRPConfig, frontendMapping *feMapping) {
+	if len(frontendMapping.backends) > 0 {
+		rpm.upsertService(config, frontendMapping)
+	} else {
+		// No backends so remove the service entry.
+		found, err := rpm.svcManager.DeleteService(frontendMapping.feAddr)
+		if !found || err != nil {
+			log.WithError(err).Debugf("Local redirect service for policy (%v)"+
+				" with frontend (%v) not deleted", config.id, frontendMapping.feAddr)
+		}
+	}
+}
+
+// deletePolicyService deletes internal state associated with the specified service.
+func (rpm *Manager) deletePolicyService(svcID k8s.ServiceID) {
+	if rp, ok := rpm.services[svcID]; ok {
+		// Get the policy config that selects this service.
+		config := rpm.policyConfigs[rp]
+		for _, m := range config.frontendMappingForSvc {
+			rpm.deletePolicyFrontend(config, m.feAddr)
+		}
+	}
+}
+
+// Upserts a service entry for the given policy config that's ready.
+func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping) {
+	frontendAddr := lb.L3n4AddrID{
+		L3n4Addr: frontendMapping.feAddr,
+		ID:       lb.ID(0),
+	}
+	var backendAddrs []lb.Backend
+	for _, be := range frontendMapping.backends {
+		backendAddrs = append(backendAddrs, lb.Backend{
+			NodeName: nodeTypes.GetName(),
+			L3n4Addr: be,
+		})
+	}
+	p := &lb.SVC{
+		Name:          config.name + localRedirectSvcStr,
+		Namespace:     config.namespace,
+		Type:          lb.SVCTypeLocalRedirect,
+		Frontend:      frontendAddr,
+		Backends:      backendAddrs,
+		TrafficPolicy: lb.SVCTrafficPolicyCluster,
+	}
+
+	if _, _, err := rpm.svcManager.UpsertService(p); err != nil {
+		log.WithError(err).Error("Error while inserting service in LB map")
+	}
```

Comment:
Couple of comments on this piece:
1. As mentioned in the previous comment, not passing the error up the stack means that the callers are unable to take action at all based on whether this fails, and they will silently assume that everything is fine. This can cause discrepancies between different portions of the system understanding the state which can be difficult to grapple with down the road.
1. I'm not sure if this error message has enough context for debugging. Do we know if `err` contains any useful information here about the related services / backends / etc?

Reply:
>not passing the error up the stack means that the callers are unable to take action at all based on whether this fails, and they will silently assume that everything is fine.

I'm not sure I follow this and your previous comments, where you mentioned something around "propagating error up the stack to user". When a user creates an LRP, the redirectpolicy manager will parse and process the LRP, thereafter other events (e.g., service/pod add/delete) are handled independently.  What exactly do you mean by "passing the error up the stack" and what actions do you propose we take here?
I checked out other places in the code, where the `UpsertService` function is called [1] [2], and they seem to have similar error handling as well, where they seem to be just logging the error.

>2. I'll add structured logging here. 

[1] https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/watcher.go#L819 
[2] https://github.com/cilium/cilium/blob/master/pkg/k8s/watchers/pod.go#L554

Reply:
My argument here is that if you have functions deep in the callstack triggering errors and you log & ignore the error, then logic higher in the callstack cannot react to these (and for instance count errors, retry, work around, track discrepancies between the desired / realized configuration, etc.). If we fail to do so and then decide in future to introduce logic to take action based on the errors then we would need to first refactor to propagate the errors correctly and then subsequently make the changes we want to make.  It's always easier to ensure error handling is introduced in the first place, all the way up the stack (where relevant), rather than refactoring all of the code later on to propagate the errors. 

In the services case of [1] it's almost at the highest point of the stack, near the watcher so there's practically only one function that the error is hidden from, meaning that there's not much logic that we could introduce to handle the error in another way. 

In the pod case of [2] it propagates the error to the caller, which is what I'm advocating for.